File: frontend/src/shared/organization/offices.ts
import type { RoleSummary } from '@/shared/api/contracts/organization/role'

export const PRESIDIUM_DEPARTMENT_SLUG = 'presidium'
export const PLATFORM_ADMIN_DEPARTMENT_SLUG = 'platform-admin'

export const DEPARTMENT_MEMBER_ROLE_KEY = 'department_member'
export const DEPARTMENT_DEPUTY_ROLE_KEY = 'department_deputy'
export const DEPARTMENT_DIRECTOR_ROLE_KEY = 'department_lead'
export const PRESIDIUM_PRESIDENT_ROLE_KEY = 'executive'
export const PRESIDIUM_SECRETARY_GENERAL_ROLE_KEY = 'presidium_secretary_general'
export const PRESIDIUM_INTERNAL_VP_ROLE_KEY = 'presidium_internal_vp'
export const PRESIDIUM_EXTERNAL_VP_ROLE_KEY = 'presidium_external_vp'
export const PLATFORM_ADMINISTRATOR_ROLE_KEY = 'platform_administrator'

export const DEPARTMENT_OFFICE_KEYS: readonly string[] = [
  DEPARTMENT_DIRECTOR_ROLE_KEY,
  DEPARTMENT_DEPUTY_ROLE_KEY,
  DEPARTMENT_MEMBER_ROLE_KEY,
]

export const PRESIDIUM_OFFICE_KEYS: readonly string[] = [
  PRESIDIUM_PRESIDENT_ROLE_KEY,
  PRESIDIUM_SECRETARY_GENERAL_ROLE_KEY,
  PRESIDIUM_INTERNAL_VP_ROLE_KEY,
  PRESIDIUM_EXTERNAL_VP_ROLE_KEY,
]

export const HIDDEN_ROLE_KEYS: readonly string[] = [
  PLATFORM_ADMINISTRATOR_ROLE_KEY,
  'department_officer',
]

export function officeKeysForDepartment(slug: string): readonly string[] {
  if (slug === PLATFORM_ADMIN_DEPARTMENT_SLUG) {
    return [PLATFORM_ADMINISTRATOR_ROLE_KEY]
  }
  if (slug === PRESIDIUM_DEPARTMENT_SLUG) {
    return PRESIDIUM_OFFICE_KEYS
  }
  return DEPARTMENT_OFFICE_KEYS
}

export function isPresidiumOffice(officeKey: string): boolean {
  return PRESIDIUM_OFFICE_KEYS.includes(officeKey)
}

export function rolesAssignableInDepartment(
  roles: readonly RoleSummary[],
  departmentSlug: string,
): RoleSummary[] {
  const allowed = new Set(officeKeysForDepartment(departmentSlug))
  return roles.filter((role) => role.key != null && allowed.has(role.key))
}

export function defaultRoleIdForDepartment(
  roles: readonly RoleSummary[],
  departmentSlug: string,
): string {
  const assignable = rolesAssignableInDepartment(roles, departmentSlug)
  const preferredKey =
    departmentSlug === PRESIDIUM_DEPARTMENT_SLUG
      ? PRESIDIUM_PRESIDENT_ROLE_KEY
      : DEPARTMENT_MEMBER_ROLE_KEY
  return assignable.find((role) => role.key === preferredKey)?.id ?? assignable[0]?.id ?? ''
}

File: backend/app/core/config/organization/offices.py
"""Office titles used for assignment and for the hidden administrator identity switcher.

Regular departments have a director, a deputy, and members. The presidium has four named seats.
The platform administrator office is never offered to anyone else.

各部门职位为部长、副部长、部员；主席团为四个具名席位。平台管理员职位不对其他人开放。
"""

from __future__ import annotations

PRESIDIUM_DEPARTMENT_SLUG = "presidium"
PLATFORM_ADMIN_DEPARTMENT_SLUG = "platform-admin"

DEPARTMENT_MEMBER_ROLE_KEY = "department_member"
DEPARTMENT_DEPUTY_ROLE_KEY = "department_deputy"
DEPARTMENT_DIRECTOR_ROLE_KEY = "department_lead"
PRESIDIUM_PRESIDENT_ROLE_KEY = "executive"
PRESIDIUM_SECRETARY_GENERAL_ROLE_KEY = "presidium_secretary_general"
PRESIDIUM_INTERNAL_VP_ROLE_KEY = "presidium_internal_vp"
PRESIDIUM_EXTERNAL_VP_ROLE_KEY = "presidium_external_vp"
PLATFORM_ADMINISTRATOR_ROLE_KEY = "platform_administrator"
ALUMNUS_ROLE_KEY = "alumnus"
RETIRED_OFFICER_ROLE_KEY = "department_officer"

DEPARTMENT_OFFICE_KEYS: tuple[str, ...] = (
    DEPARTMENT_DIRECTOR_ROLE_KEY,
    DEPARTMENT_DEPUTY_ROLE_KEY,
    DEPARTMENT_MEMBER_ROLE_KEY,
)

PRESIDIUM_OFFICE_KEYS: tuple[str, ...] = (
    PRESIDIUM_PRESIDENT_ROLE_KEY,
    PRESIDIUM_SECRETARY_GENERAL_ROLE_KEY,
    PRESIDIUM_INTERNAL_VP_ROLE_KEY,
    PRESIDIUM_EXTERNAL_VP_ROLE_KEY,
)

HIDDEN_ROLE_KEYS: frozenset[str] = frozenset(
    {PLATFORM_ADMINISTRATOR_ROLE_KEY, RETIRED_OFFICER_ROLE_KEY}
)


def office_keys_for_department(slug: str) -> tuple[str, ...]:
    """Permission-set keys that may be assigned inside one department."""
    if slug == PLATFORM_ADMIN_DEPARTMENT_SLUG:
        return (PLATFORM_ADMINISTRATOR_ROLE_KEY,)
    if slug == PRESIDIUM_DEPARTMENT_SLUG:
        return PRESIDIUM_OFFICE_KEYS
    return DEPARTMENT_OFFICE_KEYS


def build_identity_lenses() -> list[dict[str, object]]:
    """Every department-and-office pair the hidden administrator may preview."""
    from app.core.config.organization.departments import DEPARTMENT_BLUEPRINTS
    from app.core.config.organization.role_templates import ROLE_TEMPLATES_BY_KEY

    lenses: list[dict[str, object]] = []
    for blueprint in DEPARTMENT_BLUEPRINTS:
        for office_key in office_keys_for_department(blueprint.slug):
            template = ROLE_TEMPLATES_BY_KEY[office_key]
            lenses.append(
                {
                    "department_slug": blueprint.slug,
                    "department_name_en": blueprint.name_en,
                    "department_name_zh": blueprint.name_zh,
                    "office_key": template.key,
                    "office_name_en": template.name_en,
                    "office_name_zh": template.name_zh,
                    "permissions": [permission.value for permission in template.permissions],
                }
            )
    return lenses
